### PR TITLE
Project Integrations for Outgoing Webhooks

### PR DESCRIPTION
**Pull request by akitaonrails**
First version of a simple integrations administration section in the Project details. It only has Mattermost right now, but it should simple enough to add more integrations later.

The code was inserted in the existing StoryObserver that already is used to send e-mail notifications. This could be refactored into a separated service later as well.

message pushed to mattermost channel

administration of integrations on project edit form

cc @daniloisr @adbatista

### CODE REVIEW

File: app/controllers/integrations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
+class IntegrationsController < ApplicationController
+  authorize_resource
+
+  respond_to :html, :json
+
+  def index
+    @project = current_user.projects.friendly.find(params[:project_id])
```
Comment (talyssonoc) [inline app/controllers/integrations_controller.rb]: Maybe we could move that to a `before_action`?

File: app/views/notifications/started.text.erb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+<%= @owned_by.name %> has started your story '<%= @story.title %>'.
+
+<% if (@story.estimate.nil? || @story.estimate == 0) %>
```
Comment (talyssonoc) [inline app/views/notifications/started.text.erb]: Maybe just `unless @story.estimate`?

File: app/views/projects/_form.html.erb [hunk 3]
```diff
@@ -42,3 +42,37 @@
   </div>
 <% end %>
 </div>
+
+<% unless @project.new_record? %>
+  <% if @project.integrations.count > 0 %>
```
Comment (talyssonoc) [inline app/views/projects/_form.html.erb]: Can we use `@project.integrations.exists?` here?

File: app/controllers/integrations_controller.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    if @project.integrations.find_by(kind: @integration.kind)
+      flash[:alert] = "#{@integration.kind} is already configured for this project"
+    else
+      if @integration.save
+        flash[:notice] = "#{@integration.kind} was added to this project"
+      else
+        render 'index'
+        return
+      end
+    end
```
Comment (adbatista) [inline app/controllers/integrations_controller.rb]: or maybe:
```ruby
...
else
  return render :index unless @integration.save

  flash[:notice] = "#{@integration.kind} was added to this project"
end
```

File: app/models/story_observer.rb [hunk 5]
```diff
@@ -3,31 +3,52 @@ class StoryObserver < ActiveRecord::Observer
...
+            notifier = Notifications.rejected(story, story.acting_user)
+            notifier.deliver if notifier
+          end
+          IntegrationWorker.perform_async(story.project.id, "[#{story.project.name}] #{story.acting_user.name} REJECTED your story '#{story.title}'.")
         end
 
       end
 
+      # FIXME move this code to some other service that concentrates both sending email and pushing integrations
+      if notifier && story.project.integrations.count > 0
```
Comment (adbatista) [inline app/models/story_observer.rb]: why empty?

File: app/views/notifications/started.text.erb (hunk 2 above)
Comment (akitaonrails) [inline app/views/notifications/started.text.erb]: Nope, because it doesn't account for '0' (zero)

File: app/controllers/integrations_controller.rb (hunk 4 above)
Comment (akitaonrails) [inline app/controllers/integrations_controller.rb]: Too obfuscated, it's clearer the other way I think.

File: app/views/notifications/started.text.erb (hunk 2 above)
Comment (daniloisr) [inline app/views/notifications/started.text.erb]: @talyssonoc [0 is not false in Ruby](http://stackoverflow.com/q/10387515/1042324)

File: app/controllers/integrations_controller.rb [hunk 6]
```diff
@@ -0,0 +1,46 @@
+class IntegrationsController < ApplicationController
+  authorize_resource
+  before_action :load_project
+
+  respond_to :html, :json
+
+  def index
+    @integrations = @project.integrations
+    @integration = Integration.new
+    respond_with(@integrations)
```
Comment (daniloisr) [inline app/controllers/integrations_controller.rb]: `@integrations` [aren't being used on view](https://github.com/Codeminer42/cm42-central/pull/3/files#diff-2212560b4714bca5146aa0e8fc2c0558R1), so you can drop the [assign line](https://github.com/Codeminer42/cm42-central/pull/3/files#diff-93b21e526fd2ba0e475e69f6ea52b46eR8) and use just `respond_with(@project.integrations)`

File: app/controllers/integrations_controller.rb [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  end
+
+  def create
+    @integrations = @project.integrations
+    @integration = @integrations.build(kind: params[:integration][:kind]).tap do |i|
+      i.data = JSON.parse params[:integration][:data]
+    end
+
+    if @project.integrations.find_by(kind: @integration.kind)
+      flash[:alert] = "#{@integration.kind} is already configured for this project"
```
Comment (daniloisr) [inline app/controllers/integrations_controller.rb]: Why we cannot have more that one integration to the same service?

File: app/controllers/integrations_controller.rb (hunk 4 above)
Comment (adbatista) [inline app/controllers/integrations_controller.rb]: :+1:

File: app/workers/integration_worker.rb [hunk 8]
```diff
@@ -0,0 +1,16 @@
+class IntegrationWorker
+  include Sidekiq::Worker
+
+  def perform(project_id, message)
+    project = Project.find(project_id)
+    project.integrations.each do |integration|
+      if integration.kind == 'mattermost'
+        Mattermost.send(integration.data['private_uri'],
```
Comment (daniloisr) [inline app/workers/integration_worker.rb]: Later we can use STI for integrations, or use something like:

```ruby
integration.kind.contantize.send(...)
```

what you think?

File: app/mailers/notifications.rb [hunk 9]
```diff
@@ -3,28 +3,36 @@ class Notifications < ActionMailer::Base
...
+    mail :to => story.requested_by.email, :from => owned_by.email,
+      :subject => "[#{story.project.name}] Your story '#{story.title}' has been started."
+  end
+
   def delivered(story, delivered_by)
     @story = story
     @delivered_by = delivered_by
 
     mail :to => story.requested_by.email, :from => delivered_by.email,
-      :subject => "[#{story.project_name}] Your story '#{story.title}' has been delivered for acceptance."
```
Comment (adbatista) [inline app/mailers/notifications.rb]: why this change? don't story has a delegate?

Comment (akitaonrails) [inline app/mailers/notifications.rb]: I thought so, but I couldn't make it pass for some reason ('#project_name doesn't exist')

File: app/controllers/integrations_controller.rb (hunk 7 above)
Comment (akitaonrails) [inline app/controllers/integrations_controller.rb]: Why would we want to post more than once to the same service?

File: app/controllers/integrations_controller.rb [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+  def index
+    @integrations = @project.integrations
+    @integration = Integration.new
+    respond_with(@integrations)
+  end
+
+  def create
+    @integrations = @project.integrations
+    @integration = @integrations.build(kind: params[:integration][:kind]).tap do |i|
```
Comment (adbatista) [inline app/controllers/integrations_controller.rb]: `#build` accepts a block you don't need, `#tap`

File: app/controllers/integrations_controller.rb (hunk 7 above)
Comment (daniloisr) [inline app/controllers/integrations_controller.rb]: How knows ¯\\\_(ツ)\_/¯
Just asked because this validation seems to not be necessary.

File: app/mailers/notifications.rb (hunk 9 above)
Comment (adbatista) [inline app/mailers/notifications.rb]: would be easier add the delegate in story.rb, anyway :+1:
